### DISCUSSION TITLE
Import trace for requested module

### DISCUSSION THREAD

**Discussion by ullaskunder3** [vercel/next.js#49432] [Help] [answered]
### Summary

I am working on my portfolio using next.js 13 experimental [app directory], with supabase and prisma

lately I am getting unusual log after `npm run dev`

```bash
- warn ./node_modules/node-fetch/lib/index.js
Module not found: Can't resolve 'encoding' in '/run/media/dxlord/ullasDXlord/Programming-prac/vsCode/webDevelopement/WebDesign/Portfolios/final_portfolio/node_modules/node-fetch/lib'

Import trace for requested module:
./node_modules/node-fetch/lib/index.js
./node_modules/cross-fetch/dist/node-ponyfill.js
./node_modules/@supabase/supabase-js/dist/main/lib/fetch.js
./node_modules/@supabase/supabase-js/dist/main/SupabaseClient.js
./node_modules/@supabase/supabase-js/dist/main/index.js
./lib/supaClient.ts
./lib/useSession.ts
./hooks/supabaseProvider.tsx
``` 

I am still developing its working fine as expected but watching this logs is bit irritating :)

```
  "dependencies": {
    "@octokit/rest": "^19.0.7",
    "@supabase/auth-helpers-nextjs": "^0.6.1",
    "@supabase/auth-ui-react": "^0.3.6",
    "@supabase/supabase-js": "^2.21.0",
    "@supabase/ui": "^0.36.5",
    "gray-matter": "^4.0.3",
    "markdown-to-jsx": "^7.1.9",
    "next": "^13.4.1",
    "next-mdx-remote": "^4.4.1",
    "react": "^18.2.0",
    "react-dom": "^18.2.0"
  },
  "devDependencies": {
    "@tailwindcss/typography": "^0.5.9",
    "@types/node": "18.13.0",
    "@types/react": "^18.0.27",
    "autoprefixer": "^10.4.13",
    "eslint-config-next": "^13.3.0",
    "postcss": "^8.4.21",
    "tailwindcss": "^3.2.7",
    "typescript": "5.0.4"
  }
```

**Comment by roch-numbered** [comment]
I've also been running through this warning with another lib and from what I've read, the issue is due to [node-fetch](https://github.com/node-fetch/node-fetch/issues/412), [used by supabase-js](https://github.com/supabase/supabase-js/issues/612).
I think it requires that supabase updates its dependencies.

You have 2 solutions to resolve this.

## Install `encoding`

As suggested in other issues, you can install `encoding` on our own locally.

```bash
npm i -D encoding
```

## Ignore warning in `next.config.js`

If you prefer to avoid installing another (unused) lib, you can simply [ignore the warning](https://webpack.js.org/configuration/other-options/#ignorewarnings) through webpack config inside `next.config.js`

```js
webpack: config => {
  /* On `node-fetch` v2, that `supabase-js` uses,
  `encoding` package was optionally required for `.textConverted`
  which means it wasn't in `node-fetch` deps.
  See: https://github.com/node-fetch/node-fetch/issues/412.
  Since `encoding` is not part of the deps by default, when using with webpack,
  it will raise a warning message.
  This can be ignored as it doesn't prevent anything to work well. */
  config.ignoreWarnings = [
    { module: /node_modules\/node-fetch\/lib\/index\.js/ },
    { file: /node_modules\/node-fetch\/lib\/index\.js/ },
  ];

  return config;
}
```

**Reply by Digitl-Alchemyst** [reply, marked as answer]
> ```shell
> npm i -D encoding
> ```

I am having this issue in a project and it is preventing vercel deploy. This cleared up the error and i was able to proceed

**Reply by michaelo-o** [reply]
Should this be installed globally or just in the project files?

**Reply by roch-numbered** [reply]
Directly inside the project.

**Reply by NingXinHui** [reply]
What's the meaning? Does installing encoding get rid of the warning?
I do not quite understand

**Reply by ullaskunder3** [reply]
yes it does,.. and hopefully they might fix this with next patch @vercel

**Comment by roshan-gh** [comment]
Import trace for requested module:
./node_modules/vscode-languageserver-types/lib/umd/main.js
./node_modules/monaco-yaml/yaml.worker.js
./src/contexts/EditorContext.tsx
./src/app/page.tsx

I tried npm i -D encoding, but still have the warnings.

**Reply by AnonSols** [reply]
Please what's your solution, that's if you found a solution to this problem already.

**Reply by Dejnyyy** [reply]
I have exact same problem

**Comment by jakeRPowell** [comment]
I'm trying `npm run dev` on a fresh install I can't get the site running locally at all. Just getting piles of errors, ending in 

```
Import trace for requested module:
./node_modules/next/dist/shared/lib/head.js
./node_modules/next/dist/pages/_error.js
```

encoding and the next.config trick haven't worked, I've tried both the example given and 

`    { module: /node_modules\/next\/dist\/shared\/lib\/head\.js/ },
    { file: /node_modules\/next\/dist\/shared\/lib\/head\.js/ },`

**Reply by jakeRPowell** [reply]
in case it helps anyone, I found a solution somewhere saying to install a different version of node. I guess next doesn't always work with all the various dependencies etc, but it does seem to work with: Node v18.15.0

So I installed NVM, which lets you easily change between node versions, downloaded and switched to 18 and boom. Next works

**Comment by Hemdee1** [comment]
I am also running into this error whenever I'm starting my Nextjs app and the app is not starting locally.

**Reply by AnonSols** [reply]
please have you found any solution to it yet I've been on this error for days now.
On load time it shows the next landing page, after load time is completed, then I get an application error.

I've tried countless solutions on a 'Hydration error', which I get from the console of such, doesn't seem to work though.
I've tried updating my browser.
I've installed the encoding from npm.
I've downgraded my version of node to 18.15.0
I've uninstalled and installed node and vscode, honestly just had to try that as well.
I've turned off browser extension, opened the url in different browser and even tried with incognito mode as well. Nothing WORKS!!!!

Please, ANYONE how can this ever be fixed??!!

**Comment by VncntDzn** [comment]
Encountered this awhile ago but on my case it was Pubnub's react-sdk that uses Formidable, what I did was just change `node-fetch` to `formidable` and no warnings shown again.

```
Import trace for requested module:
./node_modules/formidable/src/Formidable.js
./node_modules/formidable/src/index.js
./node_modules/superagent/lib/node/index.js
./node_modules/pubnub/lib/networking/modules/web-node.js
./node_modules/pubnub/lib/node/index.js
./app/_providers/PubnubProvider.tsx
```


```

/** @type {import('next').NextConfig} */
const nextConfig = {
  images: {
    domains: ["res.cloudinary.com"],
  },
  webpack: (config) => {
    // ignore formidable warnings
    config.ignoreWarnings = [
      { module: /node_modules\/formidable\/src\/Formidable\.js/ },
      { file: /node_modules\/formidable\/src\/index\.js/ },
    ];

    return config;
  },
};

module.exports = nextConfig;
```

**Reply by AnonSols** [reply]
but why are you ignoring these warnings?

**Comment by rightson1** [comment]
Guys, i think the issue is Console Ninja extension, i disabled it an everything is back to normal

**Reply by AnonSols** [reply]
I dont even have it installed on my pc,

**Reply by rightson1** [reply]
Okay on my side, I disable console ninja , and I haven't got that error till now. You could try disabling some  extensions...

**Comment by AnonSols** [comment]
Alright after being on this error for a long time,
After trying to install the latest version of next, I installed next 13 version.
It works I can write next code but still get those Pile of errors.
thats a huge breakthrough for me because I dont get those application errors on load time and can write code on dev without the app breaking.

so my solution was to install a specific version on next which is next version 13 instead of 14

**Comment by surenachiron** [comment]
for me, I just **deleted node_module** and **changed the root folder to another folder** and it worked for me.

don't forget to run `npm I` in the new folder.

**Comment by 9a8ri3L** [comment]
It seems that, it's a webpack compiling issue, while resolving some paths, so it could be resolved manually in the ```next.config.js```.

```ts
import path from 'path';

/** @type {import('next').NextConfig} */
const nextConfig = {
  // ...
webpack: (config) => {
    config.resolve.alias['<YOUR MODULE NAME>'] = path.resolve(
      '<YOUR MODULE PATH AS SHOWN IN THE WARNING>'
    );
    return config;
  },
};

export default nextConfig;
```

Don't forget to ```return config```, otherwise, webpack config will be undefined.

**Reply by Abdi-qadir** [reply]
how i Add module name, also module path, lets help plese

**Reply by 9a8ri3L** [reply]
@Abdi-qadir 
Please post the import trace message you get, so that, we can define your module's name and path.

**Reply by Abdi-qadir** [reply]
my error is like this if some one have how to fix like this err guys
Import trace for requested module:
./node_modules/@mapbox/node-pre-gyp/lib/util/nw-pre-gyp/index.html
./node_modules/@mapbox/node-pre-gyp/lib/ sync ^\.\/.*$
./node_modules/@mapbox/node-pre-gyp/lib/node-pre-gyp.js
./node_modules/bcrypt/bcrypt.js
./pages/api/auth/[...nextauth].ts
./actions/getCurrentUser.ts
 ⨯ ./node_modules/@mapbox/node-pre-gyp/lib/util/nw-pre-gyp/index.html
Module parse failed: Unexpected token (1:0)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
> <!doctype html>
| <html>
| <head>

Import trace for requested module:
./node_modules/@mapbox/node-pre-gyp/lib/util/nw-pre-gyp/index.html
./node_modules/@mapbox/node-pre-gyp/lib/ sync ^\.\/.*$
./node_modules/@mapbox/node-pre-gyp/lib/node-pre-gyp.js
./node_modules/bcrypt/bcrypt.js
./pages/api/auth/[...nextauth].ts
./actions/getCurrentUser.ts
 ⚠ Fast Refresh had to perform a full reload due to a runtime error.
 ⚠ Fast Refresh had to perform a full reload due to a runtime error.
 GET /_next/static/webpack/0358f53ec83dcbe1.webpack.hot-update.json 500 in 119033ms
 GET /_next/static/webpack/0358f53ec83dcbe1.webpack.hot-update.json 500 in 67577ms

**Reply by 9a8ri3L** [reply]
> my error is like this if some one have how to fix like this err guys Import trace for requested module: ./node_modules/@mapbox/node-pre-gyp/lib/util/nw-pre-gyp/index.html ./node_modules/@mapbox/node-pre-gyp/lib/ sync ^./.*$ ./node_modules/@mapbox/node-pre-gyp/lib/node-pre-gyp.js ./node_modules/bcrypt/bcrypt.js ./pages/api/auth/[...nextauth].ts ./actions/getCurrentUser.ts ⨯ ./node_modules/@mapbox/node-pre-gyp/lib/util/nw-pre-gyp/index.html Module parse failed: Unexpected token (1:0) You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
> 
> > <!doctype html>
> > | 
> > |
> 
> Import trace for requested module: ./node_modules/@mapbox/node-pre-gyp/lib/util/nw-pre-gyp/index.html ./node_modules/@mapbox/node-pre-gyp/lib/ sync ^./.*$ ./node_modules/@mapbox/node-pre-gyp/lib/node-pre-gyp.js ./node_modules/bcrypt/bcrypt.js ./pages/api/auth/[...nextauth].ts ./actions/getCurrentUser.ts ⚠ Fast Refresh had to perform a full reload due to a runtime error. ⚠ Fast Refresh had to perform a full reload due to a runtime error. GET /_next/static/webpack/0358f53ec83dcbe1.webpack.hot-update.json 500 in 119033ms GET /_next/static/webpack/0358f53ec83dcbe1.webpack.hot-update.json 500 in 67577ms

`// @mapbox/node-pre-gyp`
Replace `<YOUR MODULE NAME>` with `"@mapbox/node-pre-gyp"`, and `<YOUR MODULE PATH AS SHOWN IN THE WARNING>` with  `"./node_modules/@mapbox/node-pre-gyp/lib/node-pre-gyp.js"`

`// bcrypt.js`
Replace `<YOUR MODULE NAME>` with `"bcrypt"`, and `<YOUR MODULE PATH AS SHOWN IN THE WARNING>` with  `"./node_modules/bcrypt/bcrypt.js"`


I hope this can help.
